src/linear_algebra/matrix.lean

```diff
@@ -1029,3 +1031,144 @@ lemma matrix.dot_product_eq_zero_iff {v : n → R} : (∀ w, matrix.dot_product
 ⟨λ h, matrix.dot_product_eq_zero v h, λ h w, h.symm ▸ zero_dot_product w⟩
 
 end
+
+namespace matrix
+
+variables {m n : Type u} [decidable_eq n] [fintype n] [decidable_eq m] [fintype m] {R : Type v}
+  [comm_ring R]
+
+lemma to_block_matrix_det (M : matrix m m R) (p : m → Prop) [decidable_pred p] :
+  M.det = (matrix.from_blocks (to_block M p p) (to_block M p (λ j, ¬p j))
+    (to_block M (λ j, ¬p j) p) (to_block M (λ j, ¬p j) (λ j, ¬p j))).det :=
+begin
+  rw ← matrix.det_reindex_self (equiv.sum_compl p).symm M,
+  unfold det,
+  congr, ext σ, congr, ext,
+  generalize hy : σ x = y,
+  cases x; cases y;
+  simp only [matrix.reindex_apply, to_block_apply, equiv.symm_symm,
+    equiv.sum_compl_apply_inr, equiv.sum_compl_apply_inl,
+    from_blocks_apply₁₁, from_blocks_apply₁₂, from_blocks_apply₂₁, from_blocks_apply₂₂],
+end
+
+lemma to_square_block_det (M : matrix m m R) {n : nat} (b : m → fin n) (k : fin n) :
+  (to_square_block M b k).det = (to_square_block_prop M (λ i, b i = k)).det := by simp
+
+lemma to_square_block_det' (M : matrix m m R) (b : m → ℕ) (k : ℕ) :
+  (to_square_block' M b k).det = (to_square_block_prop M (λ i, b i = k)).det := by simp
+
+example (n : ℕ) (a b : fin n) : a = b ↔ (↑a : ℕ) = ↑b := fin.ext_iff a b
+
+lemma two_block_triangular_det (M : matrix m m R) (p : m → Prop) [decidable_pred p]
+  (h : ∀ i (h1 : ¬p i) j (h2 : p j), M i j = 0) :
+  M.det = (to_square_block_prop M p).det * (to_square_block_prop M (λ i, ¬p i)).det :=
+begin
+  rw to_block_matrix_det M p,
+  convert upper_two_block_triangular_det (to_block M p p) (to_block M p (λ j, ¬p j))
+    (to_block M (λ j, ¬p j) (λ j, ¬p j)),
+  ext,
+  exact h ↑i i.2 ↑j j.2
+end
+
+lemma equiv_block_det (M : matrix m m R) {p q : m → Prop} [decidable_pred p] [decidable_pred q]
+  (e : ∀x, q x ↔ p x) : (to_square_block_prop M p).det = (to_square_block_prop M q).det :=
+by convert matrix.det_reindex_self (equiv.subtype_equiv_right e) (to_square_block_prop M q)
+
+lemma to_square_block_det'' (M : matrix m m R) {n : nat} (b : m → fin n) (k : fin n) :
+  (to_square_block M b k).det = (to_square_block' M (λ i, ↑(b i)) ↑k).det :=
+begin
+  rw [to_square_block_def', to_square_block_def],
+  apply equiv_block_det,
+  intro x,
+  apply (fin.ext_iff _ _).symm
+end
+
+/-- Let `b` map rows and columns of a square matrix `M` to `n` blocks. Then
+  `block_triangular_matrix M n b` says the matrix is block triangular. -/
```

Comment:
Since `o` is a fintype, wouldn't any `b : o -> ℕ` satisfy the condition: there is an `n` such that `∀ i, b i < n`? So we can unify `block_triangular_matrix` and `block_triangular_matrix'` by dropping the bound on `b`.

Reply:
I'll remove the `b i < n`, but we won't be able to unify the two because one is `o -> ℕ` while the other is `o -> fin n`...